File: app/app/src/androidTest/java/ru/mit/spbau/hanabi/SinglePlayerActivityTest.java
package ru.mit.spbau.hanabi;


import android.support.test.espresso.ViewInteraction;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;
import android.test.suitebuilder.annotation.LargeTest;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;

import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;
import org.hamcrest.core.IsInstanceOf;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.allOf;

@LargeTest
@RunWith(AndroidJUnit4.class)
public class SinglePlayerActivityTest {

    @Rule
    public ActivityTestRule<MainMenuActivity> mActivityTestRule = new ActivityTestRule<>(MainMenuActivity.class);

    @Test
    public void singlePlayerActivityTest() {
        ViewInteraction appCompatButton = onView(
                allOf(withId(R.id.single_player_button), withText("Single Player"),
                        childAtPosition(
                                childAtPosition(
                                        withId(android.R.id.content),
                                        0),
                                1),
                        isDisplayed()));
        appCompatButton.perform(click());

        ViewInteraction appCompatButton2 = onView(
                allOf(withId(R.id.start_button), withText("start"),
                        childAtPosition(
                                childAtPosition(
                                        withId(android.R.id.content),
                                        0),
                                3),
                        isDisplayed()));
        appCompatButton2.perform(click());

        ViewInteraction linearLayout = onView(
                allOf(childAtPosition(
                        allOf(withId(R.id.solitaire_tab_content),
                                childAtPosition(
                                        withId(android.R.id.tabcontent),
                                        0)),
                        0),
                        isDisplayed()));
        linearLayout.check(matches(isDisplayed()));

        ViewInteraction linearLayout2 = onView(
                allOf(childAtPosition(
                        allOf(withId(R.id.hands_view),
                                childAtPosition(
                                        IsInstanceOf.<View>instanceOf(android.widget.RelativeLayout.class),
                                        1)),
                        0),
                        isDisplayed()));
        linearLayout2.check(matches(isDisplayed()));

        ViewInteraction linearLayout3 = onView(
                allOf(childAtPosition(
                        allOf(withId(R.id.hands_view),
                                childAtPosition(
                                        IsInstanceOf.<View>instanceOf(android.widget.RelativeLayout.class),
                                        1)),
                        1),
                        isDisplayed()));
        linearLayout3.check(matches(isDisplayed()));

        ViewInteraction linearLayout4 = onView(
                allOf(withId(R.id.life_view),
                        childAtPosition(
                                allOf(withId(R.id.info_panel),
                                        childAtPosition(
                                                IsInstanceOf.<View>instanceOf(android.widget.LinearLayout.class),
                                                1)),
                                2),
                        isDisplayed()));
        linearLayout4.check(matches(isDisplayed()));

        ViewInteraction linearLayout5 = onView(
                allOf(withId(R.id.hints_view),
                        childAtPosition(
                                allOf(withId(R.id.info_panel),
                                        childAtPosition(
                                                IsInstanceOf.<View>instanceOf(android.widget.LinearLayout.class),
                                                1)),
                                4),
                        isDisplayed()));
        linearLayout5.check(matches(isDisplayed()));

        ViewInteraction imageView = onView(
                allOf(withId(R.id.rules_action_view),
                        childAtPosition(
                                allOf(withId(R.id.info_panel),
                                        childAtPosition(
                                                IsInstanceOf.<View>instanceOf(android.widget.LinearLayout.class),
                                                1)),
                                0),
                        isDisplayed()));
        imageView.check(matches(isDisplayed()));

        ViewInteraction linearLayout6 = onView(
                allOf(childAtPosition(
                        allOf(withId(android.R.id.tabs),
                                childAtPosition(
                                        IsInstanceOf.<View>instanceOf(android.widget.LinearLayout.class),
                                        0)),
                        1),
                        isDisplayed()));
        linearLayout6.check(matches(isDisplayed()));

    }

    private static Matcher<View> childAtPosition(
            final Matcher<View> parentMatcher, final int position) {

        return new TypeSafeMatcher<View>() {
            @Override
            public void describeTo(Description description) {
                description.appendText("Child at position " + position + " in parent ");
                parentMatcher.describeTo(description);
            }

            @Override
            public boolean matchesSafely(View view) {
                ViewParent parent = view.getParent();
                return parent instanceof ViewGroup && parentMatcher.matches(parent)
                        && view.equals(((ViewGroup) parent).getChildAt(position));
            }
        };
    }
}

File: app/app/src/androidTest/java/ru/mit/spbau/hanabi/SinglePlayerActivityTest.kt
package ru.mit.spbau.hanabi

import android.support.test.espresso.Espresso
import android.support.test.espresso.ViewAction
import android.support.test.espresso.action.ViewActions
import android.support.test.espresso.assertion.ViewAssertions
import android.support.test.espresso.matcher.ViewMatchers
import android.support.test.rule.ActivityTestRule
import org.hamcrest.CoreMatchers
import org.junit.Rule
import org.junit.Test
import android.widget.NumberPicker
import android.support.test.espresso.UiController
import android.view.View
import org.hamcrest.Matcher


class SingleGameActivityTest {
    @get:Rule
    val activityRule = ActivityTestRule(SinglePlayerActivity::class.java)

    @Test
    fun pickerAndTitleTest() {
        val title = Espresso.onView(ViewMatchers.withId(R.id.single_player_title))
        title.check(ViewAssertions.matches(ViewMatchers.withText(
                CoreMatchers.containsString("single player"))))
        val pickerText = Espresso.onView(ViewMatchers.withId(R.id.choose_players_num_text))
        pickerText.check(ViewAssertions.matches(ViewMatchers.withText(
                CoreMatchers.containsString("Choose the number of players"))))
        val picker = Espresso.onView(ViewMatchers.withId(R.id.players_number_picker))
        picker.check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
    }

    private fun startGameTest(expectedPlayersCnt: Int) {
        val startButton = Espresso.onView(ViewMatchers.withId(R.id.start_button))
        startButton.check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        startButton.perform(ViewActions.click())

        val handsView = Espresso.onView(ViewMatchers.withId(R.id.hands_view))
        handsView.check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        handsView.check(ViewAssertions.matches(ViewMatchers.hasChildCount(expectedPlayersCnt)))

        Espresso.pressBack()
    }

    private fun setNumber(num: Int): ViewAction {
        return object : ViewAction {
            override fun perform(uiController: UiController, view: View) {
                val np = view as NumberPicker
                np.value = num
            }

            override fun getDescription(): String {
                return "Set the passed number into the NumberPicker"
            }

            override fun getConstraints(): Matcher<View> {
                return ViewMatchers.isAssignableFrom(NumberPicker::class.java)
            }
        }
    }

    @Test
    fun startGameWithDifferentNumberOfPlayersTest() {
        val picker = Espresso.onView(ViewMatchers.withId(R.id.players_number_picker))
        for (i in 2..5) {
            picker.perform(setNumber(i))
            startGameTest(i)
        }
    }
}
